lib/chainsync/chainsync/analysis/calc_pnl.py

```diff
@@ -91,78 +94,40 @@ def calc_single_closeout(position: pd.Series, contract: Contract, pool_info: pd.
         except Exception as exception:  # pylint: disable=broad-except
             logging.warning("Exception caught, ignoring: %s", exception)
 
-    elif tokentype == "LP":
-        fn_args = (
-            amount,
-            min_output,
-            (  # IHyperdrive.Options
-                address,  # destination
-                True,  # asBase
-                bytes(0),  # extraData
-            ),
-        )
-        # If this fails, keep as nan and continue iterating
-        try:
-            preview_result = smart_contract_preview_transaction(
-                contract, sender, "removeLiquidity", *fn_args, block_number=position["block_number"]
-            )
-            out_pnl = Decimal(
-                preview_result["baseProceeds"]
-                # We assume all withdrawal shares are redeemable
-                + preview_result["withdrawalShares"] * pool_info["lp_share_price"].values[-1]
-            ) / Decimal(1e18)
-        except Exception as exception:  # pylint: disable=broad-except
-            logging.warning("Exception caught, ignoring: %s", exception)
-
-    elif tokentype == "WITHDRAWAL_SHARE":
-        fn_args = (
-            amount,
-            min_output,
-            (  # IHyperdrive.Options
-                address,  # destination
-                True,  # asBase
-                bytes(0),  # extraData
-            ),
-        )
-        try:
-            # For PNL, we assume all withdrawal shares are redeemable
-            # even if there are no withdrawal shares available to withdraw
-            # Hence, we don't use preview transaction here
-            out_pnl = Decimal(amount * pool_info["lp_share_price"].values[-1]) / Decimal(1e18)
-        except Exception as exception:  # pylint: disable=broad-except
-            logging.warning("Exception caught, ignoring: %s", exception)
+    elif tokentype in ["LP", "WITHDRAWAL_SHARE"]:
+        out_pnl = amount * float(lp_share_price)
+        out_pnl = Decimal(out_pnl) / Decimal(1e18)
     else:
         # Should never get here
         raise ValueError(f"Unexpected token type: {tokentype}")
     return out_pnl
 
 
 def calc_closeout_pnl(
-    current_wallet: pd.DataFrame, pool_info: pd.DataFrame, hyperdrive_contract: Contract
+    current_wallet: pd.DataFrame,
+    hyperdrive_contract: Contract,
+    hyperdrive_interface: HyperdriveReadInterface,
```

Comment:
I think it's better to still get `pool_info` here instead of `hyperdrive_interface`, and then grab the `lp_share_price` entry from that to pass to `calc_single_closeout`. The reason is that whenever we call `hyperdrive_interface.current_pool_state` it is checking the cache and possibly updating the state. We want the post-processing (basically anything in chainsync) to be able to lock _when_ it is doing whatever it is doing (like `calc_pnl`). So the better workflow is for the caller of this `calc_closeout_pnl` script to handle which pool state to use (whether it be current or at some block number) and then pass the pool info for that corresponding state.

Reply:
nice, and pool_info is already being called in data_to_analysis.py right before this, so it's free